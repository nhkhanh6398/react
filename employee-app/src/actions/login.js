import {_getUserByIdAndPassword} from "../commonUltis/_DATA"

// Initial const action
export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_AUTHED_USER = "LOGOUT_AUTHED_USER";
export const NOT_FOUND_AUTHED_USER = "NOT_FOUND_AUTHED_USER";

// Action for set authenticated user
export function setAuthedUser(login) {
    return {
        type: SET_AUTHED_USER,
        login,
    };
}

// Action to handle the case where an authenticated user cannot be found 
export function notFoundAuthedUser(login) {
    return {
        type: NOT_FOUND_AUTHED_USER,
        login,
    };
}

// Action for logout
export function logout() {
    return {
        type: LOGOUT_AUTHED_USER,
    };
}

// Process handle login account user
export function handleLogin(username, password) {
    return (dispatch) => {
        const status = localStorage.getItem('reLogin');
        
        if (status === 'false' || status === undefined || status === null) {
            localStorage.setItem('reLogin', 'true');
        }
        _getUserByIdAndPassword(username, password).then((user) => {
            if (!!user) {
                return dispatch(setAuthedUser(user));
            } else {
                return dispatch(notFoundAuthedUser(user));
            }
        })
    };
}

// Process handle logout account user
export function handleLogout() {
    return (dispatch) => {
        return dispatch(logout());
    };
}