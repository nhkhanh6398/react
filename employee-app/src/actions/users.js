// Initial const action
export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_QUESTION_USER = "ADD_QUESTION_USER";
export const ADD_ANSWER_USER = "ADD_ANSWER_USER";

// Action object for receiving a list of user
export function receiveUsers(users) {
    return {
        type: RECEIVE_USERS,
        users,
    };
}

// Action for associating a user with a answer
export function createAnswerUser(authedUser, qid, answer) {
    return {
        type: ADD_ANSWER_USER,
        authedUser,
        qid,
        answer,
    };
}

// Action for associating a user with a question
export function createQuestionUser({author, qid }) {
    return {
        type: ADD_QUESTION_USER,
        author,
        qid,
    };
}