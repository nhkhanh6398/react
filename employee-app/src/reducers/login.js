import { LOGOUT_AUTHED_USER, SET_AUTHED_USER, NOT_FOUND_AUTHED_USER } from "../actions/login";

export default function login(state = null, action) {
    switch (action.type) {
        case SET_AUTHED_USER:
            return action.login;
        case NOT_FOUND_AUTHED_USER:
            alert ('Username or password is incorrect');
            return null;
        case LOGOUT_AUTHED_USER:
            return null;
        default:
            return state;
    }
}