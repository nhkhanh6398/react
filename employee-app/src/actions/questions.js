import { createAnswerUser, createQuestionUser } from "./users";
import { _saveQuestion, _saveQuestionAnswer } from '../commonUltis/_DATA';
import { hideLoading } from "react-redux-loading-bar";

// Initial const action
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

// Action object for receiving a list of question
export function receiveQuestions(questions) {
    return {
        type: RECEIVE_QUESTIONS,
        questions,
    };
}

// Action object for adding a new question
function createQuestion(question) {
    return {
        type: ADD_QUESTION,
        question,
    };
}

// Action object for adding an answer to a question
function createAnswerQuestion(author, qid, answer) {
    return {
        type: ADD_ANSWER_QUESTION,
        author,
        qid,
        answer,
    };
}

// Process handle add questions
export function handleAddQuestion(firstOption, secondOption) {
    return (dispatch, getState) => {
        const { login } = getState();
        _saveQuestion({optionOneText: firstOption, optionTwoText: secondOption, author: login.id})
        .then((question) => {
            dispatch(createQuestion(question));
            dispatch(createQuestionUser({author: login.id, qid: question.id}));
        })
        .then(() => dispatch(hideLoading()));
    }
}

// Process handle add answers
export function handleAddAnswer(qid, answer) {
    return (dispatch, getState) => {
        const { login } = getState();
        console.log("qid: " + qid);
        console.log("answer: " + answer);
        return _saveQuestionAnswer({authedUser: login.id, qid: qid, answer: answer})
        .then(() => {
            
            dispatch(createAnswerQuestion(login.id, qid, answer));
            dispatch(createAnswerUser(login.id, qid, answer));
        })
        .then(() => dispatch(hideLoading()))
        .catch((e) => console.log("Error from handleAddAnswer: ", e));
    };
}


