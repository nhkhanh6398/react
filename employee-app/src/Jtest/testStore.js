import { createStore } from "redux";
import {persistedReducer} from '../reducers/index'
import { applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { persistStore } from 'redux-persist';

export const testStore = createStore(
    persistedReducer,
    applyMiddleware(thunk)
);

export const persistor = persistStore(testStore);