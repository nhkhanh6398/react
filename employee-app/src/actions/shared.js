import { receiveUsers } from "./users";
import { receiveQuestions } from "./questions";
import { showLoading, hideLoading } from "react-redux-loading-bar";
import {_getQuestions, _getUsers} from "../commonUltis/_DATA";

// Initial const action
export const SET_ITEM_NAV = "SET_ITEM_NAV";
export const HOME = "home";

// Process handle initial data show
export function handleInitialData() {
    return (dispatch, getState) => {
        dispatch(showLoading());
        const {users, questions, nav} = getState();
        if (Object.keys(users).length === 0 && Object.keys(questions).length === 0 && Object.keys(nav).length === 0) {
            return Promise.all([_getUsers(), _getQuestions()]).then(([users, questions]) => {
                dispatch(receiveUsers(users));
                dispatch(receiveQuestions(questions));
                dispatch({type: SET_ITEM_NAV, itemNav: HOME});
                dispatch(hideLoading());
                localStorage.setItem('reLogin', 'false');
            })
        }
    };
}