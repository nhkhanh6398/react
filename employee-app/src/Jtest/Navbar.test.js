import React from 'react';
import { render, screen } from '@testing-library/react';
import Nav from '../components/Navbar';
import { Provider } from "react-redux";
import { testStore } from "./testStore";
import { BrowserRouter } from "react-router-dom";
import {setAuthedUser} from "../actions/login";

describe("Nav", () => {
    it("Component Nav will render", () => {
        testStore.dispatch(setAuthedUser({id: "khanhnh11", password: "s"}));
        const view = render(
            <Provider store={testStore}>
                <BrowserRouter>
                    <Nav />
                </BrowserRouter>
            </Provider>
        );
        expect(view).toBeDefined();
        expect(view).toMatchSnapshot();
    });

    it("should display username of logged in user", () => {
        testStore.dispatch(setAuthedUser({id: "khanhnh11", password: "s"}));

        render(
            <Provider store={testStore}>
                <BrowserRouter>
                    <Nav/>
                </BrowserRouter>
            </Provider>
        );

        const userSpanElement = screen.getByTestId("user-information");
        expect(userSpanElement.textContent).toBe("khanhnh11");

    });
});