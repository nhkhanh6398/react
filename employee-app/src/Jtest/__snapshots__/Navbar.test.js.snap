// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Component Nav will render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand-lg navbar-light bg-success"
      >
        <button
          aria-controls="navbarTogglerDemo01"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-target="#navbarTogglerDemo01"
          data-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarTogglerDemo01"
        >
          <ul
            class="navbar-nav mr-auto mt-2 mt-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link font-weight-bold"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link font-weight-bold"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link font-weight-bold"
                href="/add"
              >
                Regist Poll
              </a>
            </li>
          </ul>
          <img
            alt="Avatar of khanhnh11"
            class="rounded-circle avatar"
            src="https://png.pngtree.com/png-vector/20190820/ourlarge/pngtree-no-avatar-vector-isolated-on-white-background-png-image_1694546.jpg"
          />
          <span
            class="nav-link"
            data-testid="user-information"
            id="user-information"
          >
            khanhnh11
          </span>
          <button
            class="btn btn-warning img-fluid"
          >
            Logout
          </button>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-lg navbar-light bg-success"
    >
      <button
        aria-controls="navbarTogglerDemo01"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-target="#navbarTogglerDemo01"
        data-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarTogglerDemo01"
      >
        <ul
          class="navbar-nav mr-auto mt-2 mt-lg-0"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link font-weight-bold"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link font-weight-bold"
              href="/leaderboard"
            >
              Leaderboard
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link font-weight-bold"
              href="/add"
            >
              Regist Poll
            </a>
          </li>
        </ul>
        <img
          alt="Avatar of khanhnh11"
          class="rounded-circle avatar"
          src="https://png.pngtree.com/png-vector/20190820/ourlarge/pngtree-no-avatar-vector-isolated-on-white-background-png-image_1694546.jpg"
        />
        <span
          class="nav-link"
          data-testid="user-information"
          id="user-information"
        >
          khanhnh11
        </span>
        <button
          class="btn btn-warning img-fluid"
        >
          Logout
        </button>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
